django/utils/functional.py

```diff
@@ -18,13 +21,30 @@ class cached_property:
     Decorator that converts a method with a single self argument into a
     property cached on the instance.
 
-    Optional ``name`` argument allows you to make cached properties of other
-    methods. (e.g.  url = cached_property(get_absolute_url, name='url') )
+    It is possible to make a cached property out of an existing method:
+    (e.g.  ``url = cached_property(get_absolute_url)`` ).
+    On Python versions before 3.6, you'll need to provide the optional ``name``
+    argument. (e.g.  ``url = cached_property(get_absolute_url, name='url')`` ).
     """
     def __init__(self, func, name=None):
+        func_name = func.__name__
+        unsupported_mangled_method = (
+            not PY36 and not name and
+            func_name.startswith('__') and not func_name.endswith('__')
+        )
+        if unsupported_mangled_method:
+            warnings.warn(
+                'cached_property does not work with mangled methods on '
+                'Python < 3.6 unless `name` is explicitly set',
+                RuntimeWarning,
+            )
         self.func = func
         self.__doc__ = getattr(func, '__doc__')
-        self.name = name or func.__name__
+        self.name = name or func_name
+
+    def __set_name__(self, owner, name):
+        if self.name == self.func.__name__:
```

Comment:
Might make sense to raise exception in this case:
```
class Test:
    @cached_property
    def a(self):
        pass
    b = a
```

Reply:
Yep